### ISSUE TITLE
Show Shared Expressions In Query Groups

### ISSUE THREAD

**Issue by samaguire** [TabularEditor/TabularEditor3#436]
It would be great if we could show the M queries in their folders under the Shared Expressions folder. i.e:

![image](https://user-images.githubusercontent.com/62320770/154919629-0451230d-7a19-4940-9fd6-a5bcfc9bdc12.png)

would be like 

![image](https://user-images.githubusercontent.com/62320770/154919760-799e9470-e4ff-4cb2-a8ad-ebbeccba5cc9.png)

**Comment by otykier** [comment]
While we could certainly display shared expressions in a folder structure according to their Query Group, my concern is that users would be confused about queries that are loaded into tables "missing" from this structure (i.e. these queries would appear as partitions on the table they are loaded into, [as discussed here](https://github.com/TabularEditor/TabularEditor3/issues/435#issuecomment-1046597504)).

I don't know what the best way to deal with this is. Maybe a new option in the TOM Explorer to display "All queries" as a separate group (instead of Shared Expressions), to better indicate that this is a different view of the model metadata, that does not follow the structure of the TOM. Any thoughts?

**Comment by samaguire** [comment]
Why not do both? Or, none? i.e. have an option to enable/disable the groups for the shared expressions (and possibly also the displaying of query groups in the partition properties) and another option to enable/disable the 'All Queries' virtual structure (which could be displayed in italic to distinguish it as virtual?). The default setting could be to have the shared expression grouping off, and the virtual folder on.

I personally do more work against a PBI dataset/model, rather than a ASSAS model, and would call myself a power user so would be comfortable with just additionally displaying the shared expressions in a folder while keeping the expression against the partition. This way I can have the queries logically grouped in a way that matches how I work: source > stage/transform > model (which would be the partitions).

However, I can understand that someone who may be less of a power user would feel more comfortable with a view that is more like PBI (the virtual view you've suggested), or I can understand that an old school ASSAS modeler who is very used to legacy connections might prefer the pure TOM layout has traditionally presented by TE.

One thing I don't think you can do away with is having/displaying the query against the partition - just thinking about advanced features like incremental refresh and how that works...

I look forward to seeing what you end up doing 😁

**Comment by otykier** [comment]
Yeah, it definitely has to be an option, as there's no one solution that will satisfy both PBI and AAS/SSAS developers. I'm thinking that we should let users choose one of the following as a preference:

- TOM (this is what we do today)
- Virtual (all Power Query expressions displayed in a virtual structure as you suggest, arranged according to their query group folders)
- Automatic (default)

The "Automatic" option would use TOM whenever an Analysis Services model is loaded (i.e. CompatibilityMode = AnalysisServices), and Virtual whenever a Power BI model is loaded.

**Comment by samaguire** [comment]
Sounds like a good middle ground 👍


Regards,
Stephen
